File: template-parts/code-library/snippets/go/retry_patterns.go
// Package retry provides Dark Factory standard retry patterns.
package retry

import (
	"context"
	"math"
	"math/rand"
	"time"
)

// Config configures retry behavior.
type Config struct {
	MaxAttempts int
	BaseDelay   time.Duration
	MaxDelay    time.Duration
	Jitter      bool  // Add randomness to delay
	OnRetry     func(attempt int, err error) // Called before each retry
}

// DefaultConfig returns a sensible default configuration.
func DefaultConfig() Config {
	return Config{
		MaxAttempts: 3,
		BaseDelay:   100 * time.Millisecond,
		MaxDelay:    5 * time.Second,
		Jitter:      true,
	}
}

// Do executes fn with retry logic according to config.
// Returns the last error if all attempts fail.
func Do(ctx context.Context, cfg Config, fn func() error) error {
	var lastErr error

	for attempt := 1; attempt <= cfg.MaxAttempts; attempt++ {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		lastErr = fn()
		if lastErr == nil {
			return nil
		}

		if attempt >= cfg.MaxAttempts {
			break
		}

		if cfg.OnRetry != nil {
			cfg.OnRetry(attempt, lastErr)
		}

		delay := calculateDelay(cfg, attempt)
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(delay):
		}
	}

	return lastErr
}

// calculateDelay computes the delay before the next retry attempt.
// Uses exponential backoff with optional jitter.
func calculateDelay(cfg Config, attempt int) time.Duration {
	delay := float64(cfg.BaseDelay) * math.Pow(2, float64(attempt-1))
	if delay > float64(cfg.MaxDelay) {
		delay = float64(cfg.MaxDelay)
	}

	if cfg.Jitter {
		// Add ±25% jitter
		jitter := delay * 0.25 * (2*rand.Float64() - 1)
		delay = delay + jitter
	}

	return time.Duration(delay)
}

// WithRetry wraps a function with retry logic.
func WithRetry(ctx context.Context, fn func() error) error {
	return Do(ctx, DefaultConfig(), fn)
}

File: template-parts/code-library/snippets/python/retry_patterns.py
"""Retry patterns with exponential backoff and jitter.

Usage:
    result = await retry(do_something, max_attempts=3, base_delay=1.0)
"""

from __future__ import annotations

import asyncio
import random
import time
from typing import TYPE_CHECKING, Any, Awaitable, Callable, TypeVar

if TYPE_CHECKING:
    pass

T = TypeVar("T")


def retry(
    func: Callable[[], T],
    *,
    max_attempts: int = 3,
    base_delay: float = 1.0,
    max_delay: float = 60.0,
    jitter: bool = True,
    exceptions: tuple[type[Exception], ...] = (Exception,),
) -> T:
    """Synchronous retry with exponential backoff.

    Args:
        func: Function to retry
        max_attempts: Maximum number of attempts
        base_delay: Base delay between retries (seconds)
        max_delay: Maximum delay (seconds)
        jitter: Add random ±25% jitter to delay
        exceptions: Tuple of exceptions to catch and retry

    Returns:
        Return value of func on success

    Raises:
        Last exception if all attempts fail
    """
    last_exc: Exception | None = None

    for attempt in range(1, max_attempts + 1):
        try:
            return func()
        except exceptions as e:
            last_exc = e
            if attempt >= max_attempts:
                break

            delay = min(base_delay * (2 ** (attempt - 1)), max_delay)
            if jitter:
                delay = delay * (0.75 + random.random() * 0.5)

            time.sleep(delay)

    if last_exc is not None:
        raise last_exc
    raise RuntimeError("retry logic error")


async def retry_async(
    coro: Callable[[], Awaitable[T]],
    *,
    max_attempts: int = 3,
    base_delay: float = 1.0,
    max_delay: float = 60.0,
    jitter: bool = True,
    exceptions: tuple[type[Exception], ...] = (Exception,),
) -> T:
    """Async retry with exponential backoff.

    Usage:
        result = await retry_async(do_something_async, max_attempts=3)
    """
    last_exc: Exception | None = None

    for attempt in range(1, max_attempts + 1):
        try:
            return await coro()
        except exceptions as e:
            last_exc = e
            if attempt >= max_attempts:
                break

            delay = min(base_delay * (2 ** (attempt - 1)), max_delay)
            if jitter:
                delay = delay * (0.75 + random.random() * 0.5)

            await asyncio.sleep(delay)

    if last_exc is not None:
        raise last_exc
    raise RuntimeError("retry_async logic error")
